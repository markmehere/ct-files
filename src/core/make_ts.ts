const fs = require("fs-extra");
const upath = require("path");

/**
 * Adds all type information to the project. If you're using an exotic include
 * (e.g. seedrandom) this isn't going to help you and you obviously have no
 * node_modules to help either. Nevertheless it should do a pretty good job of
 * defining most types. It is the callers responsibility to ensure that this is
 * only added to TypeScript projects - the types will be added regardless.
 *
 * @param proj_dir the project directory this is the directory alongside the
 *                 ict file.
 * @param core_typedefs_path the directory to ct.js core typedefs - keeping
 *                           in mind this should be the directory at runtime
 * @param lib_typedefs_path  the directory of ct.js libraries (a.k.a catmods)
 *                           - each contains a types.d.ts file (again this
 *                           should be the directory at runtime)
 * @param project the project object
 */
export async function make_ts(
    proj_dir: string,
    core_typedefs_path: string,
    lib_typedefs_path: string,
    project: any
) {
    await fs.outputFile(upath.join(proj_dir, '..', 'tsconfig.json'), JSON.stringify({
        "compilerOptions": {
          "target": "ES6",
          "module": "none",
          "noEmit": true,
          "skipLibCheck": true,
          "lib": ["DOM", "ES6"],
          "types": []
        },
        "include": [
          "./**/*.ts",
          "./**/*.d.ts"
        ],
        "exclude": [
          "node_modules"
        ]
      }, null, 2), "utf8");
    if (!(await fs.pathExists(upath.join(proj_dir, 'typedefs')))) {
        await fs.copy(core_typedefs_path, upath.join(proj_dir, 'typedefs'));
    }
    debugger;
    const promises = Object.keys(project.libs).map(async (lib) => {
        const libt_path = upath.join(lib_typedefs_path, lib, 'types.d.ts');
        if (await fs.pathExists(libt_path)) {
            if (!(await fs.pathExists(upath.join(proj_dir, 'typedefs', `${lib}.d.ts`)))) {
                await fs.copy(libt_path, upath.join(proj_dir, 'typedefs', `${lib}.d.ts`));
            }
        }
    });
    await Promise.all(promises);
}
