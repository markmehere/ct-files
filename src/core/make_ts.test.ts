import { make_ts } from "./make_ts";

jest.mock("fs-extra", () => {
    let filesystem: Record<string, string | any> = {
        "ctjs.app": {
            typedefs: { "ct.d.ts": "abc", "keyboardWorkaround.d.ts": "123", "pixi.d.ts": "456" },
            libs: { place: { "types.d.ts": "789" }, sound: { "types.d.ts": "def" }, typeless: "ghi" }
        },
        "Users": {
            "pazollim": {
                "spaceship": {
                    "spaceship.ict": "jkl",
                    "spaceship": { "asteroid.png": "mno" }
                }
            }
        }
    };
    const getNestedObject = (obj: any, pathParts: string[]) => {
        const result = pathParts.reduce((acc, part) => {
            if (acc === undefined || acc[part] === undefined) {
                return undefined;
            }
            return acc[part]
        }, obj);
        return result;
    };
    const setNestedObject = (obj: any, pathParts: string[], value: any) => {
        pathParts.reduce((acc, part, index) => {
            if (index === pathParts.length - 1) {
                acc[part] = value;
            } else {
                acc[part] = acc[part] || {};
            }
            return acc[part];
        }, obj);
    };
    return {
        pathExists: (path: string) => Promise.resolve(!!getNestedObject(filesystem, path.split('/'))),
        readFile: (path: string, _encoding: string) => {
            return Promise.resolve(getNestedObject(filesystem, path.split('/')));
        },
        outputFile: (path: string, data: string, _encoding: string) => {
            setNestedObject(filesystem, path.split('/'), data);
            return Promise.resolve();
        },
        copy: (src: string, dest: string) => {
            const srcParts = src.split('/');
            const destParts = dest.split('/');
            const srcContent = getNestedObject(filesystem, srcParts);
            if (srcContent === undefined) {
                throw new Error(`Source file not found: ${src}`);
            }
            setNestedObject(filesystem, destParts, srcContent);
            return Promise.resolve();
        },
        reset: () => {
            filesystem = {};
        }
    }
});

const fs = require("fs-extra");

describe("make_ts", () => {

    test("correctly generates TypeScript files", async () => {
        const proj_dir = "Users/pazollim/spaceship/spaceship";
        const core_typedefs_path = "ctjs.app/typedefs";
        const lib_typedefs_path = "ctjs.app/libs";
        const project = {
            libs: {
                place: { types: "place" },
                sound: { types: "sound" },
                typeless: { types: "typeless" }
            }
        };

        await make_ts(proj_dir, core_typedefs_path, lib_typedefs_path, project);

        expect(await fs.readFile('Users/pazollim/spaceship/tsconfig.json')).toMatchSnapshot();
        expect(await fs.readFile('Users/pazollim/spaceship/spaceship/typedefs')).toMatchSnapshot();
    });

})
