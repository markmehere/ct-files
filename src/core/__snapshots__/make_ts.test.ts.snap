// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`make_ts correctly generates TypeScript files 1`] = `
"{
  "compilerOptions": {
    "target": "ES6",
    "module": "none",
    "noEmit": true,
    "skipLibCheck": true,
    "lib": [
      "DOM",
      "ES6"
    ],
    "types": []
  },
  "include": [
    "./**/*.ts",
    "./**/*.d.ts"
  ],
  "exclude": [
    "node_modules"
  ]
}"
`;

exports[`make_ts correctly generates TypeScript files 2`] = `
{
  "ct.d.ts": "abc",
  "keyboardWorkaround.d.ts": "123",
  "pixi.d.ts": "456",
  "place.d.ts": "789",
  "sound.d.ts": "def",
}
`;
